File: Example/Shared/Controllers/Main/MainViewController.m
//
//  MainViewController.m
//  ThreadsDemoObjc
//
//  Created by Vitaliy Kuzmenko on 22/07/2019.
//  Copyright © 2019 Threads. All rights reserved.
//

#import "MainViewController.h"

typedef NS_ENUM(NSInteger, MainViewControllerTab) {
    MainViewControllerTabClients,
    MainViewControllerTabIntegrations,
    MainViewControllerTabUtilites,
};

@interface MainViewController () <UITabBarControllerDelegate>

@property (nonatomic, copy, nullable) NSString *unreadMessagesBadgeValue;

- (void)updateUnreadMessagesCountBadgeValue;

@end

@implementation MainViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.delegate = self;
}

- (void)setUnreadMessagesCount:(NSUInteger)unreadMessagesCount {
    self.unreadMessagesBadgeValue = unreadMessagesCount ? [NSString stringWithFormat:@"%lu", (unsigned long)unreadMessagesCount] : nil;
    [self updateUnreadMessagesCountBadgeValue];
}

- (void)updateUnreadMessagesCountBadgeValue {
    UITabBarItem *tabBarItem = self.viewControllers[MainViewControllerTabIntegrations].tabBarItem;
    UINavigationController *nc = (UINavigationController *)self.selectedViewController;
    
    if (self.selectedIndex != MainViewControllerTabIntegrations || [nc.visibleViewController conformsToProtocol:@protocol(IntegrationsProtocol)]) {
        tabBarItem.badgeValue = self.unreadMessagesBadgeValue;
    } else {
        tabBarItem.badgeValue = nil;
    }
}

#pragma mark - Show Chat from Notification

- (void)showChatForAppMarker:(NSString *)appMarker {
    self.selectedIndex = 1;
    UINavigationController *nvc = (UINavigationController *)self.viewControllers[1];
    UIViewController <IntegrationsProtocol> *vc = (UIViewController <IntegrationsProtocol> *) nvc.viewControllers[0];
    NSInteger design = [appMarker hasSuffix:@"CRG"] ? 1 : 0;
    [vc presentFromPushWithDesign:design];
}

#pragma mark - UITabBarDelegate

- (void)tabBarController:(UITabBarController *)tabBarController didSelectViewController:(UIViewController *)viewController {
    [self updateUnreadMessagesCountBadgeValue];
}

@end

File: Example/ExampleSwift/Controllers/Main/MainViewController.swift
//
//  MainViewController.swift
//  ExampleSwift
//
//  Created by Brooma Service on 24.04.2020.
//  Copyright © 2020 Threads. All rights reserved.
//

import UIKit

protocol IntegrationsProtocol {
    func presentFromPush(withDesign design: Int)
}

enum MainViewControllerTab: Int {
    case MainViewControllerTabClients = 0
    case MainViewControllerTabIntegrations = 1
    case MainViewControllerTabUtilites = 2
}

class MainViewController: UITabBarController, UITabBarControllerDelegate {
    
    var unreadMessagesBadgeValue: String?

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.delegate = self
    }
    
    func setUnreadMessagesCount(_ unreadMessagesCount: UInt) {
        
        self.unreadMessagesBadgeValue = unreadMessagesCount > 0 ? "\(unreadMessagesCount)" : nil
        self.updateUnreadMessagesCountBadgeValue()
    }

    func updateUnreadMessagesCountBadgeValue() {
        
        let tabBarItem = self.viewControllers?[MainViewControllerTab.MainViewControllerTabIntegrations.rawValue].tabBarItem
        let nc = self.selectedViewController as? UINavigationController
        
        if (self.selectedIndex != MainViewControllerTab.MainViewControllerTabIntegrations.rawValue
            || ( nc?.visibleViewController is IntegrationsProtocol)) {
            tabBarItem?.badgeValue = self.unreadMessagesBadgeValue
        } else {
            tabBarItem?.badgeValue = nil
        }
    }

    func showChat(forAppMarker appMarker: String) {
        self.selectedIndex = 1;
        let nvc = self.viewControllers?[1] as? UINavigationController
        let vc = nvc?.viewControllers[0] as? IntegrationsProtocol
        let design = appMarker.hasSuffix("CRG") ? 1 : 0;
        vc?.presentFromPush(withDesign: design)
    }

    func tabBarController(_ tabBarController : UITabBarController, didSelect didSelectViewController: UIViewController) {
        self.updateUnreadMessagesCountBadgeValue()
    }

}
